File: bindings/go/client_zalgebra.go
package kevy

import "context"

// Sorted-set algebra: ZINTERSTORE / ZUNIONSTORE / ZINTERCARD (contract
// §3.6). Available on both backends. AGGREGATE is emitted only for the
// non-default modes; WEIGHTS (when given) must be one-per-key.

// ZInterStore stores the unweighted-SUM intersection of keys into dest,
// returning dest's cardinality.
func (c *Client) ZInterStore(ctx context.Context, dest []byte, keys ...[]byte) (int64, error) {
	return c.ZInterStoreWith(ctx, dest, keys, nil, AggSum)
}

// ZInterStoreWith is the full-option intersection store.
func (c *Client) ZInterStoreWith(ctx context.Context, dest []byte, keys [][]byte, weights []float64, agg ZAggregate) (int64, error) {
	if err := checkSourceKeys(keys); err != nil {
		return 0, err
	}
	return c.execCount(ctx, zstoreArgv("ZINTERSTORE", dest, keys, weights, agg))
}

// ZUnionStore stores the unweighted-SUM union of keys into dest.
func (c *Client) ZUnionStore(ctx context.Context, dest []byte, keys ...[]byte) (int64, error) {
	return c.ZUnionStoreWith(ctx, dest, keys, nil, AggSum)
}

// ZUnionStoreWith is the full-option union store.
func (c *Client) ZUnionStoreWith(ctx context.Context, dest []byte, keys [][]byte, weights []float64, agg ZAggregate) (int64, error) {
	if err := checkSourceKeys(keys); err != nil {
		return 0, err
	}
	return c.execCount(ctx, zstoreArgv("ZUNIONSTORE", dest, keys, weights, agg))
}

// ZInterCard returns the intersection cardinality without materialising
// it. limit < 0 counts everything; limit >= 0 short-circuits at limit.
func (c *Client) ZInterCard(ctx context.Context, keys [][]byte, limit int) (int64, error) {
	if err := checkSourceKeys(keys); err != nil {
		return 0, err
	}
	argv := make([][]byte, 0, len(keys)+4)
	argv = append(argv, bs("ZINTERCARD"), itob(int64(len(keys))))
	argv = append(argv, keys...)
	if limit >= 0 {
		argv = append(argv, bs("LIMIT"), itob(int64(limit)))
	}
	return c.execCount(ctx, argv)
}

func checkSourceKeys(keys [][]byte) error {
	if len(keys) == 0 {
		return errInvalidInput("zset algebra needs at least one source key")
	}
	return nil
}

func zstoreArgv(verb string, dest []byte, keys [][]byte, weights []float64, agg ZAggregate) [][]byte {
	argv := make([][]byte, 0, len(keys)*2+6)
	argv = append(argv, bs(verb), dest, itob(int64(len(keys))))
	argv = append(argv, keys...)
	if weights != nil {
		argv = append(argv, bs("WEIGHTS"))
		for _, w := range weights {
			argv = append(argv, ftob(w))
		}
	}
	if agg != AggSum {
		argv = append(argv, bs("AGGREGATE"), agg.tag())
	}
	return argv
}

File: bindings/cpp/src/client_zalgebra.cpp
#include "argv.hpp"
#include "kevy/client.hpp"
#include "kevy/errors.hpp"
#include "reply_util.hpp"

// Sorted-set algebra: ZINTERSTORE / ZUNIONSTORE / ZINTERCARD (contract §3.6).
// Available on both backends. AGGREGATE is emitted only for non-default modes;
// WEIGHTS (when given) must be one-per-key.

namespace kevy {

using detail::Args;

namespace {

const char* agg_tag(ZAggregate a) {
  switch (a) {
    case ZAggregate::Min: return "MIN";
    case ZAggregate::Max: return "MAX";
    default: return "SUM";
  }
}

void check_source_keys(const ByteList& keys) {
  if (keys.empty()) throw InvalidInputError("zset algebra needs at least one source key");
}

Args zstore_argv(std::string_view verb, std::string_view dest, const ByteList& keys,
                 const std::optional<std::vector<double>>& weights, ZAggregate agg) {
  Args argv;
  argv.reserve(keys.size() * 2 + 6);
  argv.add(verb).add(dest).add_int(static_cast<int64_t>(keys.size())).add_all(keys);
  if (weights.has_value()) {
    argv.add("WEIGHTS");
    for (double w : *weights) argv.add_double(w);
  }
  if (agg != ZAggregate::Sum) argv.add("AGGREGATE").add(agg_tag(agg));
  return argv;
}

}  // namespace

int64_t Client::zinterstore(std::string_view dest, const ByteList& keys) {
  return zinterstore_with(dest, keys, std::nullopt, ZAggregate::Sum);
}
int64_t Client::zinterstore_with(std::string_view dest, const ByteList& keys,
                                 const std::optional<std::vector<double>>& weights, ZAggregate agg) {
  check_source_keys(keys);
  return exec_count(zstore_argv("ZINTERSTORE", dest, keys, weights, agg));
}
int64_t Client::zunionstore(std::string_view dest, const ByteList& keys) {
  return zunionstore_with(dest, keys, std::nullopt, ZAggregate::Sum);
}
int64_t Client::zunionstore_with(std::string_view dest, const ByteList& keys,
                                 const std::optional<std::vector<double>>& weights, ZAggregate agg) {
  check_source_keys(keys);
  return exec_count(zstore_argv("ZUNIONSTORE", dest, keys, weights, agg));
}

int64_t Client::zintercard(const ByteList& keys, std::optional<uint64_t> limit) {
  check_source_keys(keys);
  Args argv;
  argv.reserve(keys.size() + 4);
  argv.add("ZINTERCARD").add_int(static_cast<int64_t>(keys.size())).add_all(keys);
  if (limit.has_value()) argv.add("LIMIT").add_uint(*limit);
  return exec_count(argv);
}

}  // namespace kevy
